server/src/main/java/org/opensearch/gateway/BaseGatewayShardAllocator.java

```diff
@@ -74,7 +77,26 @@ public void allocateUnassigned(
         ExistingShardsAllocator.UnassignedAllocationHandler unassignedAllocationHandler
     ) {
         final AllocateUnassignedDecision allocateUnassignedDecision = makeAllocationDecision(shardRouting, allocation, logger);
+        executeDecision(shardRouting, allocateUnassignedDecision, allocation, unassignedAllocationHandler);
+    }
+
+    public void allocateUnassignedBatch(Set<ShardRouting> shards, RoutingAllocation allocation) {
+        // make Allocation Decisions for all shards
+        ConcurrentMap<ShardRouting, AllocateUnassignedDecision> decisionMap = makeAllocationDecision(shards, allocation, logger);
+        // get all unassigned shards
+        RoutingNodes.UnassignedShards.UnassignedIterator iterator = allocation.routingNodes().unassigned().iterator();
```

Comment:
Let's revisit this, @Gaurav614  do you want to add details about other ideas which may not require iteration of all the shards but just the shards present in the batch.

Reply:
Ya, I was thinking to add a method in `UnassignedShards` class in `RoutingNodes` like this

```
 public UnassignedIterator iteratorForShards(ListIterator<ShardRouting> shards) {
            // assert all shards unassigned

            // return object
            return new UnassignedIterator(shards);
        }
```

And a private constructor in `UnassignedIterator` class

```
            private UnassignedIterator(ListIterator<ShardRouting> iterator) {
                this.iterator = iterator;
            }
```

Now to create Unassigned iterator for given shards we do the following
```
        RoutingNodes.UnassignedShards unassigned = allocation.routingNodes().unassigned();

        List<ShardRouting> myList = new ArrayList<>(shards);
        RoutingNodes.UnassignedShards.UnassignedIterator myIterator = unassigned.iteratorForShards(myList.listIterator());
```


If it makes sense then we can go ahead with this, and I am still open for other ideas and suggestions as well.

Reply:
@Gaurav614  Updating 2 data structures is an issue here. Let's keep logic simple using map.get as ShardRouting is latest in batch (updated in GatewayAllocator).

```
if (decisionMap.containsKey(shard)) {
                        executeDecision(shard,
                            decisionMap.remove(shard),
                            allocation,
                            iterator);
                    }
```

Reply:
Resolved.